p2p/transport_test.go

```diff
@@ -0,0 +1,1016 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/tendermint/tendermint/crypto/ed25519"
+)
+
+func TestTransportMultiplexConnFilter(t *testing.T) {
+	mt := NewMultiplexTransport(
+		NodeInfo{},
+		NodeKey{
+			PrivKey: ed25519.GenPrivKey(),
+		},
+	)
+
+	MultiplexTransportConnFilters(
+		func(_ map[string]net.Conn, _ net.Conn) error { return nil },
+		func(_ map[string]net.Conn, _ net.Conn) error { return nil },
+		func(_ map[string]net.Conn, _ net.Conn) error {
+			return fmt.Errorf("rejected")
+		},
+	)(mt)
+
+	addr, err := NewNetAddressStringWithOptionalID("127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := mt.Listen(*addr); err != nil {
+		t.Fatal(err)
+	}
+
+	errc := make(chan error)
+
+	go func() {
+		addr, err := NewNetAddressStringWithOptionalID(mt.listener.Addr().String())
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		_, err = addr.Dial()
+		if err != nil {
+			errc <- err
+			return
+		}
+
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		t.Errorf("connection failed: %v", err)
+	}
+
+	_, err = mt.Accept(peerConfig{})
+	if err, ok := err.(*ErrRejected); ok {
+		if !err.IsFiltered() {
+			t.Errorf("expected peer to be filtered")
+		}
+	} else {
+		t.Errorf("expected ErrRejected")
+	}
+}
+
+func TestTransportMultiplexConnFilterTimeout(t *testing.T) {
+	mt := NewMultiplexTransport(
+		NodeInfo{},
+		NodeKey{
+			PrivKey: ed25519.GenPrivKey(),
+		},
+	)
+
+	MultiplexTransportFilterTimeout(5 * time.Millisecond)(mt)
```

Comment:
Your first option expands the surface the transport and the second will introduce a struct which will grow significantly with the options introduced and mixes different configurations. It takes the [functional option](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) (which seem to have better self-documenting character)  approach and instead of a variadic param has them as standalone calls, which are chainable. Either seems fine, I prefer the standalone version.